indexbs/indexbacked_bs.go

```diff
@@ -0,0 +1,232 @@
+	if err == nil && blk != nil {
+		return blk, nil
+	}
+
+	// load blockstore for the selected shard and try to serve the cid from that blockstore.
+	resch := make(chan dagstore.ShardResult, 1)
+	if err := ro.d.AcquireShard(ctx, sk, resch, dagstore.AcquireOpts{}); err != nil {
+		return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, err)
+	}
+	var res dagstore.ShardResult
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res = <-resch:
+		if res.Error != nil {
+			return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, res.Error)
+		}
+	}
+
+	sa := res.Accessor
+	bs, err := sa.Blockstore()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load read only blockstore for shard %s: %w", sk, err)
+	}
+
+	blk, err = bs.Get(ctx, c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	// update the block cache and the blockstore cache
+	ro.blockstoreCache.Add(sk, &accessorWithBlockstore{sa, bs})
+
+	logbs.Debugw("Get: returning after creating new blockstore", "cid", c)
+	return blk, nil
+}
+
+func (ro *IndexBackedBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	logbs.Debugw("Has called", "cid", c)
+
+	// if there is a shard that can serve the retrieval for the given cid, we have the requested cid
+	// and has should return true.
+	shards, err := ro.d.ShardsContainingMultihash(ctx, c.Hash())
+	if err != nil {
+		logbs.Debugw("Has error", "cid", c, "err", err)
+		return false, nil
+	}
+	if len(shards) == 0 {
+		logbs.Debugw("Has: returning false no error", "cid", c)
+		return false, nil
+	}
+
+	_, err = ro.shardSelectF(c, shards)
+	if err != nil && err == ErrNoShardSelected {
+		logbs.Debugw("Has error", "cid", c, "err", err)
+		return false, nil
+	}
+	if err != nil {
+		logbs.Debugw("Has error", "cid", c, "err", err)
+		return false, fmt.Errorf("failed to run shard selection function: %w", err)
+	}
+
+	logbs.Debugw("Has: returning true", "cid", c)
+	return true, nil
+}
+
+func (ro *IndexBackedBlockstore) GetSize(ctx context.Context, c cid.Cid) (int, error) {
+	logbs.Debugw("GetSize called", "cid", c)
+
+	blk, err := ro.Get(ctx, c)
+	if err != nil {
+		logbs.Debugw("GetSize error", "cid", c, "err", err)
+		return 0, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	logbs.Debugw("GetSize success", "cid", c)
+	return len(blk.RawData()), nil
+}
+
+func (ro *IndexBackedBlockstore) readFromBSCacheUnlocked(ctx context.Context, c cid.Cid, sk shard.Key) (blocks.Block, error) {
+	val, ok := ro.blockstoreCache.Get(sk)
+	if !ok {
+		return nil, ErrBlockNotFound
+	}
+
+	rbs := val.(*accessorWithBlockstore).bs
+	blk, err := rbs.Get(ctx, c)
+	if err != nil {
+		// we know that the cid we want to lookup belongs to a shard with key `sk` and
+		// so if we fail to get the corresponding block from the blockstore for that shards, something has gone wrong
+		// and we should remove the blockstore for that shard from our cache.
+		ro.blockstoreCache.Remove(sk)
+		return nil, err
+	}
```

Comment:
IIUC, whenever we need to create a new blockstore, it will first call this function and the block won't exist, so `rbs.Get(ctx, c)` will return blockstore.ErrNotFound, and it will call `blockstoreCache.Remove(sk)`

Reply:
Not sure what you mean, this method first checks if we have a blockstore for the shard and returns if we don't. Note that we've already checked that the shard has the cid/multihash we are looking for before calling this method. 

I've changed the variable naming and added a comment to make this explicit.